topic_dynamics/parsing.py

```diff
@@ -0,0 +1,150 @@
+    :param lang: the language of file.
+    :return: a list of tuples, identifier and count.
+    """
+    code = read_file(file)
+    tree = get_parser(lang).parse(code)
+    root = tree.root_node
+    identifiers = []
+    node_types = {'c': ['identifier', 'type_identifier'],
+                  'c-sharp': ['identifier', 'type_identifier'],
+                  'cpp': ['identifier', 'type_identifier'],
+                  'java': ['identifier', 'type_identifier'],
+                  'python': ['identifier', 'type_identifier']}
+
+    def traverse_tree(node: tree_sitter.Node) -> None:
+        """
+        Run down the AST from a given node and gather identifiers from its childern.
+        :param node: starting node.
+        :return: None.
+        """
+        for child in node.children:
+            if child.type in node_types[lang]:
+                start, end = get_positional_bytes(child)
+                identifier = code[start:end].decode('utf-8')
+                if '\n' not in identifier:  # Will break output files. Can add other bad characters later
+                    identifiers.append(identifier)
+            if len(child.children) != 0:
+                traverse_tree(child)
+
+    traverse_tree(root)
+    sorted_identifiers = sorted(Counter(identifiers).items(), key=itemgetter(1), reverse=True)
+
+    return sorted_identifiers
+
+
+def transform_identifiers(identifiers: List) -> List[str]:
+    """
+    Transform the original list of identifiers into the writable form.
+    :param identifiers: list of tuples, identifier and count.
+    :return: a list of identifiers in the writable for, "identifier:count".
+    """
+    formatted_identifiers = []
+    for identifier in identifiers:
+        if identifier[0].rstrip() != '':  # Checking for occurring empty tokens.
+            formatted_identifiers.append(identifier[0].rstrip() + ':' + str(identifier[1]).rstrip())
+    return formatted_identifiers
+
+
+def main(repository: str, number: int, delta: int, lang: str, output: str, output_info: str) -> None:
+    """
+    Split the repository, parse the files, write the data into a file.
+    :param repository: path to the repository to process, must have a git file.
+    :param number: the amount of dates.
+    :param delta: the time step between dates
+    :param lang: language of parsing.
+    :param output: an output file.
+    :param output_info: a file for information about output (slice indexes).
+    :return: None.
+    """
+    directory = os.path.abspath(os.path.join(repository, os.pardir, 'project_slices'))
+    os.mkdir(directory)
+    dates = get_dates(number, delta)
+    lists_of_files = {}
+    for date in dates:
+        subdirectory = os.path.abspath(os.path.join(directory, date.strftime('%Y-%m-%d')))
+        checkout_by_date(repository, subdirectory, date)
+        lists_of_files[date.strftime('%Y-%m-%d')] = get_a_list_of_files(subdirectory, get_extensions(lang))
+    indexes_of_slices = {}
```

Comment:
```suggestion
   dates_indices = {}
```

Reply:
Will do it in the next PR, because it changes the main function.